ten/one_david.py

```diff
@@ -0,0 +1,76 @@
+class BinaryHeap:
+    
+    def __init__(self, min=True):
+        self.arr = []
+        self.size = 0
+        self._sign = 2*min - 1
+        
+    def _get_val(self, ind):
+        return self.arr[ind][0]
+        
+    def push(self, val, data=None):
+        val *= self._sign
+        if self.size == len(self.arr):
+            self.arr.append((val, data))
+        else:
+            self.arr[self.size] = (val, data)
+        self.size += 1
+        
+        cur_ind = self.size - 1
+        while cur_ind > 0:
+            par_ind = cur_ind // 2
+            if self._get_val(par_ind) > self._get_val(cur_ind):
+                self.arr[par_ind], self.arr[cur_ind] = self.arr[cur_ind], self.arr[par_ind]
+                cur_ind = par_ind
+            else:
+                break
+                
+        return self
+    
+    def pop(self):
+        min_entry = (self.arr[0][0] * self._sign, self.arr[0][1])
+        self.arr[0] = self.arr[self.size - 1]
```

Comment:
If you `self.arr.pop()` here you can not keep track of size.

Reply:
yeah true